File: ta4j-core/src/main/java/org/ta4j/core/Num/DoubleNum.java
/*
  The MIT License (MIT)

  Copyright (c) 2014-2017 Marc de Verdelhan, Ta4j Organization & respective authors (see AUTHORS)

  Permission is hereby granted, free of charge, to any person obtaining a copy of
  this software and associated documentation files (the "Software"), to deal in
  the Software without restriction, including without limitation the rights to
  use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
  the Software, and to permit persons to whom the Software is furnished to do so,
  subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
  FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
  COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
  IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
  CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.Num;

import java.util.function.Function;

import static org.ta4j.core.Num.NaN.NaN;

/**
 * Representation of Double. High performance, lower precision.
 * @apiNote the delegate should never become a NaN value. No self NaN checks provided
 */
public class DoubleNum implements Num {

    private static final long serialVersionUID = -2611177221813615070L;

    private final double delegate;

    private final static double EPS = 0.00001; // precision

    @Override
    public Function<Number, Num> function() {
        return DoubleNum::valueOf;
    }

    public DoubleNum(double val){
        delegate = val;
    }


    @Override
    public Double getDelegate() {
        return delegate;
    }

    @Override
    public String getName() {
        return "DoubleNum";
    }

    @Override
    public Num plus(Num augend) {
        if (augend == NaN){
            return NaN;
        }
        if (!(augend instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format("The underlying data types '%s' and '%s' are not the same",this.getClass(), augend.getClass()));
        }

        return new DoubleNum(delegate+((DoubleNum) augend).delegate);
    }

    @Override
    public Num minus(Num subtrahend) {
        if (subtrahend == NaN){
            return NaN;
        }
        if (!(subtrahend instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), subtrahend.getClass()));
        }

        return new DoubleNum(delegate-((DoubleNum) subtrahend).delegate);
    }

    @Override
    public Num multipliedBy(Num multiplicand) {
        if (multiplicand == NaN){
            return NaN;
        }
        if (!(multiplicand instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), multiplicand.getClass()));
        }

        return new DoubleNum(delegate*((DoubleNum) multiplicand).delegate);
    }

    @Override
    public Num dividedBy(Num divisor) {
        if (divisor == NaN){
            return NaN;
        }
        if (!(divisor instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), divisor.getClass()));
        }
        DoubleNum divisorD = (DoubleNum) divisor;
        if(((DoubleNum) divisor).delegate==0){
            return NaN;
        }
        return new DoubleNum(delegate/divisorD.delegate);
    }

    @Override
    public Num remainder(Num divisor) {
        if (divisor == NaN){
            return NaN;
        }
        if (!(divisor instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), divisor.getClass()));
        }

        return new DoubleNum(delegate%((DoubleNum) divisor).delegate);
    }

    @Override
    public Num pow(int n) {
        return new DoubleNum(Math.pow(delegate,n));
    }

    @Override
    public Num abs() {
        return new DoubleNum(Math.abs(delegate));
    }

    @Override
    public boolean isZero() {
        return delegate == 0;
    }

    @Override
    public boolean isPositive() {
        return delegate > 0;
    }

    @Override
    public boolean isPositiveOrZero() {
        return delegate >= 0;
    }

    @Override
    public boolean isNegative() {
        return delegate < 0;
    }

    @Override
    public boolean isNegativeOrZero() {
        return delegate <= 0;
    }

    @Override
    public boolean isEqual(Num other) {
        if (other == NaN){
            return true;
        }
        if (!(other instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), other.getClass()));
        }
        return delegate == ((DoubleNum) other).delegate;
    }

    /**
     * Checks if this value is greater than another.
     * @param other the other value, not null
     * @return true is this is greater than the specified value, false otherwise
     */
    public boolean isGreaterThan(Num other) {
        return (other != NaN) && compareTo(other) > 0;
    }

    /**
     * Checks if this value is greater than or equal to another.
     * @param other the other value, not null
     * @return true is this is greater than or equal to the specified value, false otherwise
     */
    public boolean isGreaterThanOrEqual(Num other) {
        return (other != NaN) && compareTo(other) > -1;
    }

    /**
     * Checks if this value is less than another.
     * @param other the other value, not null
     * @return true is this is less than the specified value, false otherwise
     */
    public boolean isLessThan(Num other) {
        return (other != NaN) && compareTo(other) < 0;
    }

    @Override
    public boolean isLessThanOrEqual(Num other) {
        return (other != NaN) && compareTo(other) < 1;
    }

    /**
     * Checks if this value is less than or equal to another.
     * @param other the other value, not null
     * @return true is this is less than or equal to the specified value, false otherwise
     */
    public boolean isLessThanOrEqual(Number other) {
        return (other != NaN) && compareTo(function().apply(other)) < 1;
    }

    @Override
    public Num min(Num other) {
        if (other == NaN){
            return NaN;
        }
        if (!(other instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), other.getClass()));
        }
        return new DoubleNum(Math.min(delegate,((DoubleNum) other).delegate));
    }

    @Override
    public Num max(Num other) {
        if (other == NaN){
            return NaN;
        }
        if (!(other instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), other.getClass()));
        }
        return new DoubleNum(Math.max(delegate,((DoubleNum)other).delegate));
    }

    @Override
    public boolean isNaN() {
        return this.equals(NaN);
    }


    @Override
    public int hashCode() {
        return ((Double) (delegate)).hashCode();
    }

    @Override
    public String toString() {
        return Double.toString(delegate);

    }

    @Override
    public boolean equals(Object obj) {
        if(obj==null){
            return false;
        }
        if (this == NaN && obj == NaN){
            return true;
        }
        if(!(obj instanceof DoubleNum)) {
            return false;
        }
        DoubleNum doubleNumObj = (DoubleNum) obj;

        return Math.abs(delegate - doubleNumObj.delegate) < EPS;
    }

    @Override
    public int compareTo(Num o) {
        if (this == NaN || o == NaN){
            return 0;
        }
        if (!(o instanceof DoubleNum)){
            throw new IllegalArgumentException(String.format(
                    "The underlying data types '%s' and '%s' are not the same"
                    ,this.getClass(), o.getClass()));
        }
        DoubleNum doubleNumO = (DoubleNum)o;
        return Double.compare(delegate, doubleNumO.delegate);
    }

    public static Num valueOf(int i) {
        return new DoubleNum((double)i);
    }

    public static Num valueOf(long i) {
        return new DoubleNum((double)i);
    }

    public static Num valueOf(short i) {
        return new DoubleNum((double)i);
    }

    public static Num valueOf(float i) {
        return new DoubleNum((double)i);
    }

    public static Num valueOf(String i) {
        return new DoubleNum(Double.parseDouble(i));
    }

    public static Num valueOf(Number i) {
        return new DoubleNum(Double.parseDouble(i.toString()));
    }
}

File: ta4j-core/src/main/kotlin/org/ta4j/core/num/DoubleNum.kt
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2017-2022 Ta4j Organization & respective
 * authors (see AUTHORS)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of
 * this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to
 * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
 * the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
package org.ta4j.core.num

import java.util.function.Function
import kotlin.math.ln
import kotlin.math.pow

/**
 * Representation of Double. High performance, lower precision.
 *
 * @apiNote the delegate should never become a NaN value. No self NaN checks
 * provided
 */
class DoubleNum private constructor(override val delegate: Double) : Num {
    override fun zero(): Num {
        return ZERO
    }

    override fun one(): Num {
        return ONE
    }

    override fun hundred(): Num {
        return HUNDRED
    }

    override fun function(): Function<Number?, Num> {
        return Function { i: Number? -> valueOf(i) }
    }

    override val name: String
        get() = this.javaClass.simpleName

    override fun plus(augend: Num): Num {
        return if (augend.isNaN) NaN.NaN else DoubleNum(delegate + (augend as DoubleNum).delegate)
    }

    override fun minus(subtrahend: Num): Num {
        return if (subtrahend.isNaN) NaN.NaN else DoubleNum(delegate - (subtrahend as DoubleNum).delegate)
    }

    override fun multipliedBy(multiplicand: Num): Num {
        return if (multiplicand.isNaN) NaN.NaN else DoubleNum(delegate * (multiplicand as DoubleNum).delegate)
    }

    override fun dividedBy(divisor: Num): Num {
        if (divisor.isNaN || divisor.isZero) {
            return NaN.NaN
        }
        val divisorD = divisor as DoubleNum
        return DoubleNum(delegate / divisorD.delegate)
    }

    override fun remainder(divisor: Num): Num {
        return if (divisor.isNaN) NaN.NaN else DoubleNum(delegate % (divisor as DoubleNum).delegate)
    }

    override fun floor(): Num {
        return DoubleNum(kotlin.math.floor(delegate))
    }

    override fun ceil(): Num {
        return DoubleNum(kotlin.math.ceil(delegate))
    }

    override fun pow(n: Int): Num {
        return DoubleNum(delegate.pow(n.toDouble()))
    }

    override fun pow(n: Num): Num {
        return DoubleNum(delegate.pow(n.doubleValue()))
    }

    override fun sqrt(): Num {
        return if (delegate < 0) {
            NaN.NaN
        } else DoubleNum(kotlin.math.sqrt(delegate))
    }

    override fun sqrt(precision: Int): Num {
        return sqrt()
    }

    override fun abs(): Num {
        return DoubleNum(kotlin.math.abs(delegate))
    }

    override fun negate(): Num {
        return DoubleNum(-delegate)
    }

    override val isZero: Boolean
        get() = delegate == 0.0
    override val isPositive: Boolean
        get() = delegate > 0
    override val isPositiveOrZero: Boolean
        get() = delegate >= 0
    override val isNegative: Boolean
        get() = delegate < 0
    override val isNegativeOrZero: Boolean
        get() = delegate <= 0

    override fun isEqual(other: Num?): Boolean {
        if (other==null) return false
        return !other.isNaN && delegate == (other as DoubleNum?)!!.delegate
    }

    override fun log(): Num {
        return if (delegate <= 0) {
            NaN.NaN
        } else DoubleNum(ln(delegate))
    }

    /**
     * Checks if this value is greater than another.
     *
     * @param other the other value, not null
     * @return true is this is greater than the specified value, false otherwise
     */
    override fun isGreaterThan(other: Num?): Boolean {
        if (other==null) return true
        return !other.isNaN && compareTo(other) > 0
    }

    /**
     * Checks if this value is greater than or equal to another.
     *
     * @param other the other value, not null
     * @return true is this is greater than or equal to the specified value, false
     * otherwise
     */
    override fun isGreaterThanOrEqual(other: Num?): Boolean {
        if (other==null) return true
        return !other.isNaN && compareTo(other) > -1
    }

    /**
     * Checks if this value is less than another.
     *
     * @param other the other value, not null
     * @return true is this is less than the specified value, false otherwise
     */
    override fun isLessThan(other: Num?): Boolean {
        if (other==null) return true
        return !other.isNaN && compareTo(other) < 0
    }

    override fun isLessThanOrEqual(other: Num?): Boolean {
        if (other==null) return true
        return !other.isNaN && compareTo(other) < 1
    }

    override fun min(other: Num): Num {
        return if (other.isNaN) NaN.NaN else DoubleNum(kotlin.math.min(delegate, (other as DoubleNum).delegate))
    }

    override fun max(other: Num): Num {
        return if (other.isNaN) NaN.NaN else DoubleNum(kotlin.math.max(delegate, (other as DoubleNum).delegate))
    }

    override fun hashCode(): Int {
        return delegate.hashCode()
    }

    override fun toString(): String {
        return delegate.toString()
    }

    override fun equals(other: Any?): Boolean {
        if (other !is DoubleNum) {
            return false
        }
        return kotlin.math.abs(delegate - other.delegate) < EPS
    }

    override fun compareTo(other: Num): Int {
        if (this === NaN.NaN || other === NaN.NaN) {
            return 0
        }
        val doubleNumO = other as DoubleNum
        return delegate.compareTo(doubleNumO.delegate)
    }

    companion object {
        private val ZERO = valueOf(0)
        private val ONE = valueOf(1)
        private val HUNDRED = valueOf(100)
        private const val EPS = 0.00001 // precision

        @JvmStatic
        fun valueOf(i: Int): DoubleNum {
            return DoubleNum(i.toDouble())
        }

        @JvmStatic
        fun valueOf(i: Long): DoubleNum {
            return DoubleNum(i.toDouble())
        }

        @JvmStatic
        fun valueOf(i: Short): DoubleNum {
            return DoubleNum(i.toDouble())
        }

        @JvmStatic
        fun valueOf(i: Float): DoubleNum {
            return DoubleNum(i.toDouble())
        }

        @JvmStatic
        fun valueOf(i: String): DoubleNum {
            return DoubleNum(i.toDouble())
        }

        @JvmStatic
        fun valueOf(i: Number?): DoubleNum {
            return DoubleNum(i!!.toDouble())
        }
    }
}